### PR TITLE
Adding support for printing to work with a shared volume (cloud environment)

### PR DESCRIPTION
**Pull request by tiago-s-vieira-alb**
Before this change, print-lib generates a file into disk, and cache the file pointer into a Java map (in memory).
The problem of this, is after generating a file, the next request to get the file MUST be to the same geoserver instance. Other geoservers, doesn't have any reference to this generated file. In a k8s environment this is a problem.

Before implementation:
- Not cloud environment and not shared volume, the loadbalancer must have sticky session --> it works
- Not cloud environment and shared volume, the loadbalancer must have sticky session --> it works
- Not cloud environment and shared volume, if loadbalancer do not have sticky session --> it will crash when get from a different geoserver then the one that generates the file
- Cloud environment, the loadbalancer cannot have sticky session --> it will crash when get from a different geoserver then the one that generates the file

To solve it:
- When in printing phase, instead of java map, we persist the information into a JSON file on disk (next to the PDF file)
- When requesting the file, the code now reads the JSON file from the disk and always provides the response
- The purge of old files (more than 10 minutes) was maintained

IN SUMMARY:
What was saved in an instance into an hashmap, now is persisted on disk.

After releasing version with this PR, create a ticket in geoserver like this "https://osgeo-org.atlassian.net/browse/GEOS-11480" to update mapfish-print version.

### CODE REVIEW

**Comment by tiago-s-vieira-alb** [comment]
Hi @jodygarnett , can you please review PR?
It will be great it could be integrated into a next minor geoserver version.
Thanks

**Comment by jodygarnett** [comment]
Okay! Yeah many other geoserver extensions end up with a shared database (rather than a shared json) to communicate state across instances. 

I wish we had a good shared module for that which could use an appropriate storage / communication to avoid all the duplication we are seeing.

**Comment by jodygarnett** [comment]
Can I ask you to include a note on the docs about this as it is a pretty important changed :)

**Comment by tiago-s-vieira-alb** [comment]
> Can I ask you to include a note on the docs about this as it is a pretty important changed :)

@ritika-t-thakur-alb updated documentation.

File: docs/upgrade.rst [hunk 1]
```diff
@@ -4,7 +4,7 @@ Upgrade
 Version 2.3.2
 -------------
 
-* Support for printing in a cloud environment and in an environment without sticky sessions : generates a requested printout metadata json file and stores it in the tempDir. With this information peristed, there is no need for sticky sessions and printing module can be used in a cloud environment with shared volume.
+* Support for printing in a cloud environment and in an environment without sticky sessions : generates a requested printout metadata json file and stores it in the tempDir. With this information persisted, there is no need for sticky sessions and printing module can be used in a cloud environment with shared volume.
```
Comment (jodygarnett) [inline docs/upgrade.rst:7]: So how does this work in a could environment - don’t they need to share a temp folder or something?

Comment (ritika-t-thakur-alb) [reply docs/upgrade.rst:7]: Hello @jodygarnett , 
I elaborated my documentation a bit more. 
Answering your question: depending on the cloud environment, engineers an pick an option to share a volume between multiple instances and pass the path in the existing MAPFISH_PDF_FOLDER environment variable.
I have already updated installation.rst file with the documentation for MAPFISH_PDF_FOLDER.

I think the steps for creating a shared volume is not related with this PR.

Comment (tiago-s-vieira-alb) [reply docs/upgrade.rst:7]: Hi @jodygarnett, i think everything is done from our side.
Do you need more information?

**Comment by jodygarnett** [comment]
If you rebase to resolve conflict we should be able to merge?

**Comment by tiago-s-vieira-alb** [comment]
Hi @jodygarnett i think we don't have any conflict.
Can you check please?

**Comment by jodygarnett** [comment]
Okay I could not rebase but I can do a squash and merge

**Comment by tiago-s-vieira-alb** [comment]
@jodygarnett when you release print-2.3.2, please update also in geoserver:
https://osgeo-org.atlassian.net/browse/GEOS-11480

Thank you :)

**Comment by jodygarnett** [comment]
I think it is all done
